src/pymedphys/dicom/_level3/dicom_collection.py

```diff
@@ -0,0 +1,135 @@
+# Copyright (C) 2019 Simon Biggs, Matt Jennings
+
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as published
+# by the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version (the "AGPL-3.0+").
+
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU Affero General Public License and the additional terms for more
+# details.
+
+# You should have received a copy of the GNU Affero General Public License
+# along with this program. If not, see <http://www.gnu.org/licenses/>.
+
+# ADDITIONAL TERMS are also included as allowed by Section 7 of the GNU
+# Affero General Public License. These additional terms are Sections 1, 5,
+# 6, 7, 8, and 9 from the Apache License, Version 2.0 (the "Apache-2.0")
+# where all references to the definition "License" are instead defined to
+# mean the AGPL-3.0+.
+
+# You should have received a copy of the Apache-2.0 along with this
+# program. If not, see <http://www.apache.org/licenses/LICENSE-2.0>.
+
+
+import numpy as np
+import pydicom
+
+from .._level1.create import dicom_dataset_from_dict
+from .._level2.anonymise import anonymise_dicom_dataset
+from .._level2.dose import (
+    extract_dicom_patient_xyz, convert_xyz_to_dicom_coords)
+
+from ...libutils import get_imports
+
+IMPORTS = get_imports(globals())
+
+# pylint: disable=W0201
+
+
+class DicomBase:
+    def __init__(self, dataset):
+        self.dataset = dataset
+
+    @classmethod
+    def from_file(cls, filepath):
+        dataset = pydicom.read_file(filepath, force=True)
+
+        return cls(dataset)
+
+    @classmethod
+    def from_dict(cls, dictionary):
+        dataset = dicom_dataset_from_dict(dictionary)
+
+        return cls(dataset)
+
+    def __repr__(self):
+        return self.dataset.__repr__()
+
+    def save_as(self, filepath):
+        self.dataset.save_as(filepath)
+
+    def anonymise(self):
+        anonymised = anonymise_dicom_dataset(self.dataset)
+
+        return self.__class__(anonymised)
+
+
+class DicomDose(DicomBase):
+
+    def __init__(self, dataset):
+        super().__init__(dataset)
+
+        self.mask = None
+
+    @property
+    def values(self) -> np.ndarray:
+        return self.dataset.pixel_array * self.dataset.DoseGridScaling
+
+    @property
+    def units(self):
+        return self.dataset.DoseUnits
+
+    @units.setter
+    def units(self, units_used):
+        self.dataset.DoseUnits = units_used
+
+    # TODO: Need to refactor the following to still be a view to the dataset
+    # but not needlessly call the entire function.
+    @property
+    def x(self):
+        x_value, _, _ = extract_dicom_patient_xyz(self.dataset)
+        return x_value
```

Comment:
Because I am no longer protecting the `dataset` structure with a defensive copy, need to be very sure to only have pointers to that `dataset` object instead of pulling the data out. Otherwise honest changes to the `dataset` would result in an internally inconsistent object.

This way of implementing views is not ideal. I think we should consider inverting the dependency here.

The logic for the each individual coord should belong within the object, and then `extract_dicom_patient_xyz` should depend on this object and just be a helper wrapper function.

What do you think @Centrus007?

Reply:
This is actually a complicated problem.

Of course, *one* way around it is to instantiate DicomBase with a **filepath** rather than a pydicom dataset (like I had it originally, hehe).

Reply:
Although, if we encourage the use of `from_file()` in examples, docs, etc. I think this problem is largely circumvented.

Reply:
Yeah, I do think `self.dataset` ought to be a property. That will help.

Reply:
Sorry for the barrage, I was typing as I was thinking. I just want to add the following for clarity:

> Otherwise honest changes to the dataset would result in an internally inconsistent object.

I don't think we should permit changes to `self.dataset` at all. I think we should enforce (if possible) or at least strongly hint that a new DicomXXX object should be created if we want to make changes to its underlying pydicom dataset

Reply:
> In this current iteration a change to dataset here is just fine. x, y, z, coords, values, units will all continue to mimic what's in dataset even if it is changed by the user.

Yes I know, your comment seemed to be asking how to *avoid* the current implementation, which does feels ugly/inefficient (having to call extraction functions upon every call to, say, `self.x`).

My thinking was that simply reading DicomXXX attributes will be a more common use-case than *changing* them. If so, my thoughts were to return a new DicomXXX instance for any change required. Or at least to *copy* and replace `self.dataset` and run any dependent attribute-setting methods. Maybe that is the worse of the two trade offs...

> I envisage having a machine ID property for example on DicomPlan, it can be accessed, and changed either via the dataset or via the object property.

I really don't like this optionality. It's safe (for the moment), but unintuitive in my opinion. I do think that the average programmer would look at this:

```python
ds = pydicom.dcmread(ds_filepath)
dose = DicomDose(ds)
```
and assume that `dose` and `ds` are now logically separate entities. Note that this does not mean dose has to be *immutable*, just that any changes are best done via `dose.dataset = x`, and not have `ds = x` alter `dose.dataset`.

Its very debatable, but that's my thinking at the moment.

Reply:
The above comment notwithstanding, I think I'm happy to merge :)

Reply:
I agree. Hmm. What about a separate trade off, have an optional `no_copy` parameter that defaults to false?

Reply:
Actually not a bad idea. Incidentally, Numpy itself does this for its arrays. `numpy.array()` has a `copy` parameter that defaults to True. See https://docs.scipy.org/doc/numpy/reference/generated/numpy.array.html

Reply:
Awesome, well let's fall in line with numpy standard and have also have a copy parameter that defaults to True?

Reply:
Agreed.

Shall we merge this in and implement the copy later? Or squeeze it in now?